### PR TITLE
Add Unprocessable Entity Exception as in #2106

### PR DESCRIPTION
**Pull request by janjitsu**
Continue the line of tought in #2106

I want to be able to send 422 Http Error to inform form validation errors on my RESTful Api

### CODE REVIEW

**Comment by samdark** [comment]
Are you sure it's correct code for form validation errors?

File: framework/web/UnprocessableEntityHttpException.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * exception with status code 422.
+ *
+ * Use this exception to inform that the server understands the content type of
+ * the request entity and the syntax of that request entity is correct but was 
+ * unable to process the contained instructions. For example, to return form 
+ * validation errors. 
+ *
+ * @link http://www.webdav.org/specs/rfc2518.html#STATUS_422
+ * @author Jan Silva <janfrs3@gmail.com>
+ * @since 2.0
```
Comment (Renkas) [inline framework/web/UnprocessableEntityHttpException.php]: I'm pretty surethis Exception is not available in version 2.0 of Yii. Should change to the real release version - 2.0.7 probably

**Comment by Renkas** [comment]
AFAIK there is no correct "Validation failed" HTTP response code defined and this 422 is commonly used for that.

**Comment by samdark** [comment]
I saw 400 used for the purpose.

**Comment by Renkas** [comment]
400 would be fine if the data sent to server is syntactically wrong eg. malformed XML/json etc. But if the syntax of sent data is right then the proper response would still be 422.

**Comment by janfrs** [comment]
I agree with @Renkas, but I also agree with @samdark, AFAIK there's an ongoing debate to what is the more appropriate error code to send on form validation errors, as you can see [here](http://stackoverflow.com/questions/16133923/400-vs-422-response-to-post-of-data), [here](http://www.bennadel.com/blog/2434-http-status-codes-for-invalid-data-400-vs-422.htm) and [here](http://stackoverflow.com/questions/1959947/whats-an-appropriate-http-status-code-to-return-by-a-rest-api-service-for-a-val) 

I, personally, want my API to respond differently when the client is sending all expected fields, but some information on them isn't right. For that I find 422 (UnprocessableEntity) a more suitable response.

File: framework/CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,7 @@ Yii Framework 2 Change Log
 2.0.7 under development
 -----------------------
 
+- Enh #2106: Added Unprocessable Entity Http Exception (janfrs)
```
Comment (samdark) [inline framework/CHANGELOG.md]: Should be grouped with other `Enh` and ordered according to its number.

File: framework/web/UnprocessableEntityHttpException.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+namespace yii\web;
+
+/**
+ * UnprocessableEntityHttpException represents an "Unprocessable Entity" HTTP 
+ * exception with status code 422.
+ *
+ * Use this exception to inform that the server understands the content type of
+ * the request entity and the syntax of that request entity is correct but was
```
Comment (samdark) [inline framework/web/UnprocessableEntityHttpException.php]: correct but was → correct but **server** was

**Comment by samdark** [comment]
I'm OK adding it.

**Comment by samdark** [comment]
Thanks!

**Comment by cebe** [comment]
FYI: https://github.com/yiisoft/yii2/commit/73d6f32234057cab4d097773f303e0fd8c0388b4

**Pull request merged**